File: Algorithm/C++/904.fruit-into-baskets.cpp
 /*
 * @lc app=leetcode id=904 lang=cpp
 *
 * [904] Fruit Into Baskets
 */

// @lc code=start
#include <vector>
#include <unordered_map>

class Solution {
public:
    int totalFruit(std::vector<int>& fruits) {
        int left = 0, right = 0, result = 0, valid_fruit = 0;
        std::unordered_map<int,int> counter;
        while (right <= fruits.size()) {
            while (valid_fruit > 2) {
                counter[fruits[left]]--;
                if (counter[fruits[left]] == 0) {
                    valid_fruit--;
                }
                left++;
            }
            result = std::max(result, right - left);
            if (right < fruits.size()) {
                if (counter.find(fruits[right]) == counter.end() ||
                    counter[fruits[right]] == 0) {
                        valid_fruit++;
                    }
                counter[fruits[right]]++;
            }
            right++;
        }
        return result;
    }
};
// @lc code=end

File: Algorithm/Python/904.fruit-into-baskets.py
#
# @lc app=leetcode id=904 lang=python
#
# [904] Fruit Into Baskets
#

# @lc code=start

from collections import Counter

class Solution(object):
    def totalFruit(self, fruits):
        """
        :type fruits: List[int]
        :rtype: int
        """
        counter = Counter()
        left, right, result, valid_fruit = 0, 0, 0, 0
        while right <= len(fruits):
            while valid_fruit > 2:
                counter[fruits[left]] -= 1
                if counter[fruits[left]] == 0:
                    valid_fruit -= 1
                left += 1
            result = max(result, right - left)
            if right < len(fruits):
                if counter[fruits[right]] == 0:
                    valid_fruit += 1
                counter[fruits[right]] += 1
            right += 1
        return result
        
# @lc code=end

